arbnode/batch_poster.go

```diff
@@ -1143,17 +1143,9 @@ func (b *BatchPoster) maybePostSequencerBatch(ctx context.Context) (bool, error)
 		if nonce != gotNonce || !bytes.Equal(batchPositionBytes, gotMeta) {
 			return false, fmt.Errorf("%w: nonce changed from %d to %d while creating batch", storage.ErrStorageRace, nonce, gotNonce)
 		}
-
-		cert, err := b.daWriter.Store(ctx, sequencerMsg, uint64(time.Now().Add(config.DASRetentionPeriod).Unix()), []byte{}) // b.daWriter will append signature if enabled
-		if errors.Is(err, das.BatchToDasFailed) {
-			if config.DisableDasFallbackStoreDataOnChain {
-				return false, errors.New("unable to batch to DAS and fallback storing data on chain is disabled")
-			}
-			log.Warn("Falling back to storing data on chain", "err", err)
-		} else if err != nil {
+		sequencerMsg, err = b.dapWriter.Store(ctx, sequencerMsg, uint64(time.Now().Add(config.DASRetentionPeriod).Unix()), []byte{}, config.DisableDapFallbackStoreDataOnChain)
```

Comment:
It seems unbalanced that `daprovider.Writer` is used only to write to the DAS and `daprovider.Reader` is used to read both DAS and blobs. I'm not sure how you'd structure it to be symmetric, though, since on one hand posting a batch to the DAS is a separate operation to posting the batch itself to L1, and with blobs both get posted together with the L1 tx.

Additionally for blob batch posting there is some unique logic around deciding whether or not to post as a 4844 tx or not https://github.com/OffchainLabs/nitro/pull/2158 which is different to the logic around whether to post a DAS batch or not (falls back on error only), so that also would make it hard to unify.

This is just an observation, we should seek more feedback about this from others on the team.

Reply:
You are right, I could not find a way to wrap blob posting to follow a common writer interface that other DAs implement.